pkg/reconciler/trustedresources/verify.go

```diff
@@ -50,3 +61,145 @@ func verifyInterface(
+			APIVersion: "tekton.dev/v1beta1",
+			Kind:       "Pipeline"},
+		ObjectMeta: pm,
+		Spec:       pipelineObj.PipelineSpec(),
+	}
+	verifiers, err := getVerifiers(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, verifier := range verifiers {
+		if err := VerifyInterface(ctx, pipeline, verifier, signature); err == nil {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("Pipeline %v in namespace %v fails verification", pipeline.Name, pipeline.Namespace)
+}
+
+// prepareObjectMeta will filter the objectmeta and extract the signature.
+func prepareObjectMeta(in metav1.ObjectMeta) (metav1.ObjectMeta, []byte, error) {
+	out := metav1.ObjectMeta{}
+
+	// exclude the fields populated by system.
+	out.Name = in.Name
+	out.GenerateName = in.GenerateName
+	out.Namespace = in.Namespace
+
+	if in.Labels != nil {
+		out.Labels = make(map[string]string)
+		for k, v := range in.Labels {
+			out.Labels[k] = v
+		}
+	}
+
+	out.Annotations = make(map[string]string)
+	for k, v := range in.Annotations {
+		out.Annotations[k] = v
+	}
+
+	// exclude the annotations added by other components
+	// Task annotations are unlikely to be changed, we need to make sure other components
+	// like resolver doesn't modify the annotations, otherwise the verification will fail
+	delete(out.Annotations, "kubectl-client-side-apply")
+	delete(out.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
+
+	// signature should be contained in annotation
+	sig, ok := in.Annotations[SignatureAnnotation]
+	if !ok {
+		return out, nil, fmt.Errorf("signature is missing")
+	}
+	// extract signature
+	signature, err := base64.StdEncoding.DecodeString(sig)
+	if err != nil {
+		return out, nil, err
+	}
+	delete(out.Annotations, SignatureAnnotation)
+
+	return out, signature, nil
+}
+
+// getVerifiers get all verifiers from configmap
+func getVerifiers(ctx context.Context) ([]signature.Verifier, error) {
+	// Check if kms is configured, if not check if cosign key configured.
+	cfg := config.FromContextOrDefaults(ctx)
+	verifiers := []signature.Verifier{}
+
+	for key := range cfg.TrustedResources.Keys {
+		v, err := loadPublicKey(ctx, key)
+		if err == nil {
+			verifiers = append(verifiers, v)
+		}
+	}
+
+	if len(verifiers) == 0 {
+		return verifiers, fmt.Errorf("no public keys are founded for verification")
+	}
+
+	return verifiers, nil
+}
+
+// loadPublicKey is a wrapper for VerifierForKeyRef, hardcoding SHA256 as the hash algorithm
+func loadPublicKey(ctx context.Context, keyRef string) (verifier signature.Verifier, err error) {
+	return verifierForKeyRef(ctx, keyRef, crypto.SHA256)
+}
+
+// verifierForKeyRef parses the given keyRef, loads the key and returns an appropriate
+// verifier using the provided hash algorithm
+func verifierForKeyRef(ctx context.Context, keyRef string, hashAlgorithm crypto.Hash) (verifier signature.Verifier, err error) {
+	// The key could be plaintext, in a file, at a URL, or in KMS.
+	if kmsKey, err := kms.Get(ctx, keyRef, hashAlgorithm); err == nil {
```

Comment:
It is not obvious from docs that url is allowed as a key. Could you add that as well in the doc.

Reply:
KMS url reference not supported in this PR due to kms dependency not added. I'm planning to add them in next PR. Maybe I should remove this and add back in next PR?